packages/flutter_tools/lib/src/build_system/build_system.dart

```diff
@@ -0,0 +1,581 @@
+    for (Target target in ordered) {
+      final List<FileSystemEntity> inputs = target.resolveInputs(environment);
+      final _InvocationEvaluation evaluation = target._canSkipInvocation(inputs, environment);
+
+      if (evaluation.canSkip) {
+        printTrace('Skipping target: ${target.name}');
+        continue;
+      }
+      printTrace('${target.name}: Starting');
+      await target.invocation(inputs, environment);
+
+      printTrace('${target.name}: Complete');
+      final List<FileSystemEntity> outputs = target.resolveOutputs(environment);
+      target.writeStamp(inputs, outputs, environment, evaluation.shas);
+    }
+  }
+
+  /// Describe the target `name` and all of its dependencies.
+  List<Map<String, Object>> describe(
+    String name,
+    Environment environment,
+  ) {
+    // Compute the required order of targets.
+    final List<Target> ordered = _computeTargetOrder(targets, name, environment);
+    final List<Map<String, Object>> result = <Map<String, Object>>[];
+    for (Target target in ordered) {
+      result.add(target.toJson(environment));
+    }
+    return result;
+  }
+
+  static List<Target> _computeTargetOrder(List<Target> allTargets, String name, Environment environment) {
+    /// Step 1: Find the target we have been asked to invoke.
+    final Target target = allTargets.firstWhere((Target target) => target.name  == name, orElse: () => null);
+
+    /// Error case 1: The target name did not exist. In the interest of brevity,
+    /// we do nothing, but to be complete we should print an error which fully
+    /// describes the error along with targets that are closest in spelling.
+    /// Furthermore, we should log this error in analytics to track if it is
+    /// being hit frequently.
+    if (target == null) {
+      throw Exception('No registered target named $name.');
+    }
+
+    // Step 2: Flatten the dependencies into a set of targets to invoke
+    // in order. This is done performing a depth-first search and adding
+    // a target to the resulting order if it either has no dependencies,
+    // or if all dependencies have already been added. We take advantage
+    // of the fact that the literal [Set] provides an ordered [LinkedHashSet].
+    final Set<Target> ordered = <Target>{};
+    final Set<Target> visited = <Target>{};
+    void orderTargets(Target current) {
+      // If we've already visited this target, returning so we do not
+      // repeat ourselves in the resulting order
+      if (visited.contains(current)) {
+        return;
+      }
+      visited.add(current);
+      // If the target has no dependencies, we can add it the current order.
+      if (current.dependencies.isEmpty) {
+        ordered.add(current);
+        return;
+      }
+      // Visit each dependency recursively, and when complete add this
+      // target to the order.
+      current.dependencies.forEach(orderTargets);
+      // We can't hit the case where we visit a target recursively, right?
```

Comment:
Visiting the graph to check for cycles before doing anything else is probably a good idea. Then you can assume a DAG in all other operations on the graph.